dev-docs/RFCs/v8.0/ui-module-rfc.md

```diff
@@ -0,0 +1,137 @@
+  setProps: (props: PropsT) => void;
+
+  /// Event hooks
+  onViewStateChange?: () => void;
+  onRedraw?: () => void;
+  onHover?: (info) => void;
+  onClick?: (info) => void;
+}
+```
+
+2. Add a new module `@deck.gl/ui` which exports common UI components that implement the `UIControl` interface.
+
+3. Add the following APIs to the `Deck` class:
+
+```js
+/// Prop type for React / declarative usage
+new Deck({
+  uiControls: UIControl[]
+})
+
+/// Imperative methods for vanilla JS usage
+addUIControl(control: UIControl): void
+removeUIControl(control: UIControl): void
+```
+
+Users will use this new module like the following:
+
+```js
+import {NavigationControl, FullscreenControl} from '@deck.gl/ui'
+
+new Deck({
+  ...
+  uiControls: [
+    new NavigationControl(),
+    new FullscreenControl()
+  ]
+})
+```
+
+## Implementation details and concerns
+
+### Terminology
+
+React and Mapbox users use "controls" to refer to UI components. However, we already call an input handler "controller" in the deck.gl context. I have seen users not differentiating the two in issues and discussions. I am concerned that introducing another concept called "control" is going to add to the confusion.
+
+### React
+
+All components in `@deck.gl/ui` should work without React. `@deck.gl/react` can offer thin wrappers that work with the `DeckGL` component:
+
+```jsx
+import {NavigationControl, FullscreenControl} from '@deck.gl/react'
+
+<DeckGL>
+  <NavigationControl />
+  <FullscreenControl />
+</DeckGL>
+```
+
+### Performance impact
+
+A new `UIControlManager` will be added to `Deck` alongside `ViewManager` `EffectManager` etc. It will handle calling the lifecycle and event hooks of registered controls. For existing apps, where no UI controls are used, the performance impact is projected to be neglegible.
+
+I would like to call out the expected behavior when the `uiControls` prop shallowly changes. Following existing behavior pattern of `layers`, `views` and `effects`, upon `setProps`:
+
+- `UIControlManager` deeply compares the elements of `uiControls`' old and new values. Two controls are be the "same" if they share the same constructor and the same id.
+- If a control only exists in the new array, call `onAdd`
+- If a control only exists in the old array, call `onRemove`
+- If a control exists in both old and new arrays, transfer the new props to the old instance, instead of adopting the new instance
+
+This would avoid the performance hit by constantly adding/removing controls when using React.
+
+### CSS packaging
+
+react-map-gl (inherited from mapbox-gl) requires a separate CSS stylesheet be included in the page's head. deck.gl traditionally uses CSS-in-JS (see [getTooltip](/docs/api-reference/core/deck.md)). While I like the convenience of everything-in-one-bundle approach, the new UI module will have much more complex DOM structures, and it will be harder to design an API that allows users to customize the look and feel of their UI.
```

Comment:
> more complex DOM structures

Along these lines, how have you imagined UI be implemented such that it's easy to maintain, expressive, but also minimal? Would we employ react (or preact), HTML templates, or some other UI-in-JS framework to help? While a framework is possibly overkill, defining UI in vanilla JS can get messy and fragile.

I also acknowledge users can implement their own control as they please, but we'll be setting the example implementation.

I could imagine each instance of a UI controller mounting their own react-dom onto their container.

Reply:
Yes, agreed, At the same time having to deal with the separate mapbox CSS is a pain, just an extra failure/frustration point for users. 

I suppose it is possible to add a <style> element programmatically and have that be a long multiline string representing a proper style sheet.

Reply:
Kepler prolifically added classes to their DOM elements to allow for customization, and that works pretty well. It's stable, just a bit verbose.

It'd be nice to utilize CSS custom properties for easier theming too https://developer.mozilla.org/en-US/docs/Web/CSS/Using_CSS_custom_properties

These choices would allow us to separate end-user customization from implementation since the widget could be built with any CSS system as long as classes and properties were also used.